network/gossip/libp2p/middleware.go

```diff
@@ -194,6 +190,16 @@ func (m *Middleware) Start(ov middleware.Overlay) error {
 		return fmt.Errorf("failed to start libp2p node: %w", err)
 	}
 
+	libp2pConnector, err := NewLibp2pConnector(m.libP2PNode.libP2PHost)
+	if err != nil {
+		return fmt.Errorf("failed to create libp2pConnector: %w", err)
+	}
+	m.peerManager = NewPeerManager(m.ctx, m.log, m.topology, libp2pConnector)
+	err = m.peerManager.Start()
+	if err != nil {
+		return fmt.Errorf("failed to start peer manager: %w", err)
+	}
+
```

Comment:
When either of the above error cases happens, we have already started a LibP2P node instance, which should be stopped gracefully. Currently, we delegate this to the application layer who calls `net.Ready` and should call `net.Done` when the former returns an error. Not sure if it is a safe mechanism to rely on a graceful stop of the network to the caller upon an internal error. Maybe an atomic consideration makes more sense, i.e., either both LibP2P node and peer manager start, or none of them.

Reply:
ic ...I think the right way of doing that would be pass the context to `Start` instead of `NewMiddleware`. All libp2p based components are based on that context. Then if there is an error in start, the context can be cancelled thus cancelling everything that depends on it. I will try it out.